# Reto Integrador - Unidad 1

# PARTE 1. NumPy & Fundamentos Python

## 1.1. Setup del Entorno
Para iniciar con el reto, inicialmnere se crea un entorno virtual llamado reto_u1, se Instala numpy, pandas y sqlalchemy


## - Se realiza la limpieza del archivo
Se creó una copia del archivo CSV para corregir errores de formato que impedían su lectura, en este caso las comillas ("") y la tilde (´) del apellido López. El archivo original se conservó sin modificaciones y la limpieza se realizó sobre la copia para continuar con el análisis.

## - Se importan las librerías

In [7]:
import numpy as np
import pandas as pd

## - Se cargan los datos desde csv

In [8]:
#Cargamos csv limpio
df_ventas = pd.read_csv(
    "../data/ventas_techventas_limpio.csv",
    encoding="latin-1"
)

df_ventas.head(61)

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
0,1001,2024-01-05,C001,TechCorp SA,Norte,Laptop Pro 15,Electronica,2,1200.0,0.10,Ana Lopez
1,1002,2024-01-07,C002,Innovatek,Sur,Mouse Inalambrico,Perifericos,15,25.0,0.00,Carlos Ruiz
2,1003,2024-01-10,C003,DataSolutions,Centro,Monitor 27,Electronica,5,350.0,0.05,Ana Lopez
3,1004,2024-01-12,C001,TechCorp SA,Norte,Teclado Mecanico,Perifericos,10,85.0,0.10,Luis Mora
4,1005,2024-01-15,C004,CloudNet,Este,Laptop Pro 15,Electronica,3,1200.0,0.15,Carlos Ruiz
5,1006,2024-01-18,C005,Sistemas Globales,Oeste,SSD 1TB,Almacenamiento,20,95.0,0.00,Maria Torres
6,1007,2024-01-20,C002,Innovatek,Sur,Monitor 27,Electronica,2,350.0,0.05,Ana Lopez
7,1008,2024-01-22,C006,NetPrime,Norte,Router WiFi 6,Redes,8,120.0,0.00,Luis Mora
8,1009,2024-01-25,C003,DataSolutions,Centro,SSD 1TB,Almacenamiento,12,95.0,0.10,Maria Torres
9,1010,2024-01-28,C007,AlphaTech,Sur,Laptop Pro 15,Electronica,1,1200.0,0.00,Carlos Ruiz


## Análisis y exploración inicial del dataset

In [9]:
#Verificar los tipos de datos

df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_id         60 non-null     int64  
 1   fecha            60 non-null     str    
 2   cliente_id       60 non-null     str    
 3   cliente_nombre   60 non-null     str    
 4   region           60 non-null     str    
 5   producto         59 non-null     str    
 6   categoria        60 non-null     str    
 7   cantidad         60 non-null     int64  
 8   precio_unitario  60 non-null     float64
 9   descuento        60 non-null     float64
 10  vendedor         60 non-null     str    
dtypes: float64(2), int64(2), str(7)
memory usage: 5.3 KB


In [12]:
df_ventas.dtypes

order_id             int64
fecha                  str
cliente_id             str
cliente_nombre         str
region                 str
producto               str
categoria              str
cantidad             int64
precio_unitario    float64
descuento          float64
vendedor               str
dtype: object

Análisis Durante la exploración de los datos se identificó un registro con una cantidad negativa, lo que evidencia la necesidad de realizar una etapa de limpieza de datos antes de efectuar análisis más avanzados

In [ ]:
df.isna().sum()

order_id           0
fecha              0
cliente_id         0
cliente_nombre     0
region             0
producto           1
categoria          0
cantidad           0
precio_unitario    0
descuento          0
vendedor           0
dtype: int64

# 1.2. Algebra Lineal Aplicada

#  - Creación de arreglos con NumPy

In [ ]:
cantidades = df["cantidad"].to_numpy()

precios = df["precio_unitario"].to_numpy()

print(cantidades)

print(precios)

[ 2 15  5 10  3 20  2  8 12  1  4 30  5  3 25  2  6 18 12  2  4 30  8  3
  4 20 15  6  1 10  3 22  5 25  2  4 12 18  2  3 15  5  2  3 20  6 18 -1
  4 25  3  2 20 35  5  2  7  9 20  4]
[1200.   25.  350.   85. 1200.   95.  350.  120.   95. 1200.  350.   25.
   85.  120.   95. 1200.  350.   95.   25. 1200.  120.   95.   85.  350.
 1200.   25.   95.  120. 1200.   85.  350.   95. 1200.   25.  120.  350.
   85.   95.  350. 1200.   25.  120.  350. 1200.   95.   85.   25. 1200.
  120.   95.  350. 1200.   25.   95.  350. 1200.  120.   85.   95.  350.]


## - Se Verifica el tipo de dato

In [ ]:
type(cantidades)

numpy.ndarray

## - Producto punto

In [ ]:
ingreso_total = np.dot(cantidades, precios)

print("Ingreso total:", ingreso_total)

Ingreso total: 88205.0


## - Multiplicación elemento a elemento

In [ ]:
ingresos = cantidades * precios

print(ingresos)

[ 2400.   375.  1750.   850.  3600.  1900.   700.   960.  1140.  1200.
  1400.   750.   425.   360.  2375.  2400.  2100.  1710.   300.  2400.
   480.  2850.   680.  1050.  4800.   500.  1425.   720.  1200.   850.
  1050.  2090.  6000.   625.   240.  1400.  1020.  1710.   700.  3600.
   375.   600.   700.  3600.  1900.   510.   450. -1200.   480.  2375.
  1050.  2400.   500.  3325.  1750.  2400.   840.   765.  1900.  1400.]


In [ ]:
cantidades @ precios

np.float64(88205.0)

## Comparación de operaciones

La multiplicación elemento a elemento calcula el ingreso generado por cada venta individual al multiplicar la cantidad por el precio unitario.

El producto punto (`np.dot`) realiza todas esas multiplicaciones y posteriormente suma los resultados, obteniendo el ingreso bruto total de las ventas, que en este caso es **88 205.0**.

# 1.3. Estadísticas con NumPy

In [ ]:
media = np.mean(precios)
mediana = np.median(precios)
desviacion = np.std(precios)
percentiles = np.percentile(precios, [25, 75])

print("Media:", media)
print("Mediana:", mediana)
print("Desviación estándar:", desviacion)
print("Percentiles:", percentiles)

Media: 378.25
Mediana: 120.0
Desviación estándar: 445.0597572236789
Percentiles: [ 92.5 350. ]


## Interpretación de los resultados

- La media del precio unitario es de **378.25**, lo que representa el precio promedio de los productos vendidos.

- La mediana es de **120.00**, indicando que la mitad de los productos tiene un precio menor o igual a este valor.

- La desviación estándar muestra una alta dispersión en los precios debido a que existen productos económicos, como el mouse inalámbrico, y otros mucho más costosos, como el Laptop Pro 15.

- El percentil 25 indica que el 25 % de los productos tiene un precio menor o igual a 92.5, mientras que el percentil 75 muestra que el 75 % de los productos tiene un precio menor o igual a 350.

- Se observa un ingreso negativo debido a un registro con cantidad igual a -1. Este tipo de inconsistencias será corregido en la etapa de preparación de datos utilizando Pandas.

# 1.4. Reflexión sobre la metodología CRISP-DM

Durante el desarrollo de esta actividad entendí que antes de realizar cualquier análisis es importante conocer y comprender los datos con los que se va a trabajar. Al principio pensé que solo era cargar el archivo y hacer los cálculos, pero al revisar el dataset encontré algunos errores de formato y registros con inconsistencias, por lo que fue necesario preparar la información antes de continuar.

Con el uso de NumPy pude realizar operaciones matemáticas y obtener estadísticas que ayudan a entender mejor el comportamiento de las ventas. Esto me permitió ver que un buen análisis no depende únicamente de saber programar, sino también de interpretar los resultados y detectar posibles problemas en los datos.

Considero que la metodología CRISP-DM es una guía útil porque organiza el proceso de análisis de datos paso a paso. En este ejercicio comprendí que cada etapa es importante y que dedicar tiempo a preparar correctamente la información facilita el trabajo en las fases siguientes y permite obtener resultados más confiables.
